### PR TITLE
Implement generic invokeWithinContext for all editor panes

### PR DESCRIPTION
**Pull request by roblourens**
Continues PR #104694
and #104894

Breaking this part out to simplify things. I don't fully understand the hierarchy of editor types, but I think this should catch the cases that we have currently. BaseTextEditor can do the code editor check and delegate to the existing invokeWithinContext method for those, TextDiffEditor can override it for its deal, and I fixed the EditorService to use the new generic method.

Looking at other usages of the code editor's `invokeWithinContext`, it looks like they are all only concerned with code editors and so it would be ok to let them call it at that level. But @bpasero did you want to change those usages more aggressively?

To handle @eamodio's requirement properly, we also need to change the way that the Left/Right diff focus context keys work, so that they are set at the editor's top-level ContextKeyService and change in value, rather than being defined as constants in the individual left/right ContextKeyServices, because we need the CKS to trigger a change event that the menu is listening to.

### CODE REVIEW

**Review by eamodio — APPROVED** [review]
LGTM!

File: src/vs/workbench/browser/parts/editor/baseEditor.ts [hunk 1]
```diff
@@ -83,6 +84,13 @@ export abstract class BaseEditor extends Composite implements IEditorPane {
 	 */
 	protected abstract createEditor(parent: HTMLElement): void;
 
+	/**
+	 * Should be overridden by editors that have their own ScopedContextKeyService
+	 */
+	invokeWithinContext<T>(fn: (accessor: ServicesAccessor) => T): T | null {
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/editor/baseEditor.ts:90]: I think this is dangerous because the calling side will assume that when `null` is returned, the editor did not invoke in the context, but it could well be that a function that was passed in to invoke returns `null`. So a better approach imho is to pass in a function that is being called as fallback if no implementation exists? 

Alternatively we could make this method optional `?` and the caller has to probe if this method exists.

File: src/vs/workbench/services/editor/browser/editorService.ts [hunk 2]
```diff
@@ -783,9 +783,9 @@ export class EditorService extends Disposable implements EditorServiceImpl {
 	//#region invokeWithinEditorContext()
 
 	invokeWithinEditorContext<T>(fn: (accessor: ServicesAccessor) => T): T {
-		const activeTextEditorControl = this.activeTextEditorControl;
-		if (isCodeEditor(activeTextEditorControl)) {
-			return activeTextEditorControl.invokeWithinContext(fn);
+		const result = this.activeEditorPane?.invokeWithinContext(fn);
```
Comment (bpasero) [inline src/vs/workbench/services/editor/browser/editorService.ts:786]: We have to preserve the semantics of the code how it used to be: check the implementation of `activeTextEditorControl` around `isCompositeEditor`:

```ts
get activeTextEditorControl(): ICodeEditor | IDiffEditor | undefined {
	const activeEditorPane = this.activeEditorPane;
	if (activeEditorPane) {
		const activeControl = activeEditorPane.getControl();
		if (isCodeEditor(activeControl) || isDiffEditor(activeControl)) {
			return activeControl;
		}
		if (isCompositeEditor(activeControl) && isCodeEditor(activeControl.activeCodeEditor)) {
			return activeControl.activeCodeEditor;
		}
	}

	return undefined;
}
```

I think @jrieken added code specifically to support notebooks here. Are you planning to implement this method properly in the notebook editor then to preserve the semantics?

File: src/vs/workbench/browser/parts/editor/textDiffEditor.ts [hunk 3]
```diff
@@ -60,6 +60,19 @@ export class TextDiffEditor extends BaseTextEditor implements ITextDiffEditorPan
 		this.doSaveOrClearTextDiffEditorViewState(editor);
 	}
 
+	invokeWithinContext<T>(fn: (accessor: ServicesAccessor) => T): T | null {
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/editor/textDiffEditor.ts:63]: I understand this fixes issues for @eamodio where the editor title actions are not aware of the focused side of the diff editor. But isn't this also risky to change? Here is an example:
* I install an extension that allows to open a file as binary data (hex editor)
* this extension adds to any editor and upon click opens the hex editor
* I open a diff editor from SCM
* I click on the left hand side (the right hand side is editable, left hand is not)
* I click on the Hex action
* We now treat the left hand side as active editor
* the hex action opens the binary form of the left hand editor which is just a representation of the SCM state, but not the state on disk.

In other words, this may break a lot of actions that so far have assumed the context is always the right hand diff editor. This is especially bad because we have no good indication which side of the diff editor has focus actually.

**Review by bpasero — CHANGES_REQUESTED** [review]

File: src/vs/workbench/browser/parts/editor/textDiffEditor.ts [hunk 4]
```diff
@@ -60,6 +60,19 @@ export class TextDiffEditor extends BaseTextEditor implements ITextDiffEditorPan
...
+		const control = this.getControl();
+		if (!control) {
+			return null;
+		}
+
+		if (control.getOriginalEditor().hasTextFocus()) {
+			return control.getOriginalEditor().invokeWithinContext(fn);
+		} else {
+			return control.getModifiedEditor().invokeWithinContext(fn);
+		}
```
Comment (walidbarakat) [inline src/vs/workbench/browser/parts/editor/textDiffEditor.ts:73]: since you used ternary if condition in invokeWithinContext template implementation in textEditor.ts, it's possible to use the same style here as well: 
```suggestion
		return (control.getOriginalEditor().hasTextFocus())?
                      control.getOriginalEditor().invokeWithinContext(fn):
                      control.getModifiedEditor().invokeWithinContext(fn);
```

**Review by walidbarakat — CHANGES_REQUESTED** [review]

File: src/vs/workbench/browser/parts/editor/baseEditor.ts (hunk 1 above)
Comment (roblourens) [reply src/vs/workbench/browser/parts/editor/baseEditor.ts:90]: Yeah, I was wondering whether that is an issue since "no implementation" is basically the same as "implementation but no service", but it's better to be explicit. I like making the method optional.

File: src/vs/workbench/services/editor/browser/editorService.ts (hunk 2 above)
Comment (roblourens) [reply src/vs/workbench/services/editor/browser/editorService.ts:786]: Ah, I think I didn't fully understand the composite editor concept - I think that if TitleControl was aware of composite editors, then it could get the ContextKeyService out of the editor in the notebook and we wouldn't have needed any of these changes.

So I guess the question is should this work by default for all composite editors or should every composite editor have to implement this itself.

File: src/vs/workbench/browser/parts/editor/textDiffEditor.ts (hunk 3 above)
Comment (roblourens) [reply src/vs/workbench/browser/parts/editor/textDiffEditor.ts:63]: Would the resource that it opens come from the context? Or would it be getting that some other way?

File: src/vs/workbench/services/editor/browser/editorService.ts (hunk 2 above)
Comment (bpasero) [reply src/vs/workbench/services/editor/browser/editorService.ts:786]: I suggest you coordinate with Jo on this topic, he added the concept.

File: src/vs/workbench/browser/parts/editor/textDiffEditor.ts (hunk 3 above)
Comment (bpasero) [reply src/vs/workbench/browser/parts/editor/textDiffEditor.ts:63]: Yeah good point, I think it actually could come in from the context we set here:

https://github.com/microsoft/vscode/blob/d50e84208f0358b4ee17b32c5394ea5605641b81/src/vs/workbench/browser/parts/editor/titleControl.ts#L232

And then is passed as static argument to the actions here:

https://github.com/microsoft/vscode/blob/d50e84208f0358b4ee17b32c5394ea5605641b81/src/vs/workbench/browser/parts/editor/titleControl.ts#L248

So it would mean that the context is always the primary (right hand side) editor but the context key service might operate on the left hand side.

File: src/vs/workbench/services/editor/browser/editorService.ts (hunk 2 above)
Comment (roblourens) [reply src/vs/workbench/services/editor/browser/editorService.ts:786]: I think that we can either

- Allow any editor to implement/override invokeWithinContext
  - And in this option, need to decide whether it works by default for composite editors or if they still have to implement it (i don't know what other composite editors there are)
- Or leave it as a back door only for text editors, making TitleControl aware of composite editors so that it works for notebooks. In this case I guess we don't have the option to override it for custom/other editors

Comment (jrieken) [reply src/vs/workbench/services/editor/browser/editorService.ts:786]: The composite editor is simply a list of code editor of which one might be active. Having the code above ensures that the command palette works for the correct cell editor, but there aren't so many users of `invokeWithEditorContext` so this could be easily changed. 

On a related note: it seems that all users of `invokeWithinEditorContext` are only interested in the "relevant" context key service. So would it maybe be better to simply expose that and leave `ICodeEditorWidget#invokeWith" as an implementation detail?

Comment (roblourens) [reply src/vs/workbench/services/editor/browser/editorService.ts:786]: > it seems that all users of invokeWithinEditorContext are only interested in the "relevant" context key service

@bpasero what do you think about this, what else do you expect this to be used for? As-is, it opens up editor internals much more than necessary.

Comment (bpasero) [reply src/vs/workbench/services/editor/browser/editorService.ts:786]: Yeah worth a shot. I do not have any stakes in `invokeWithinEditorContext` so any solution is fine with me that does not break existing behaviour.

**Comment by roblourens** [comment]
I think this is ready to go, but I don't really have a reason to merge this now an hour before code freeze, so I will click merge on Friday, get NotebookEditor updated in https://github.com/microsoft/vscode/pull/104694 for this, then it will finally be done.

File: src/vs/workbench/browser/parts/editor/editorGroupView.ts [hunk 5]
```diff
@@ -861,8 +861,8 @@ export class EditorGroupView extends Themable implements IEditorGroupView {
 		}
 	}
 
-	invokeWithinContext<T>(fn: (accessor: ServicesAccessor) => T): T {
-		return this.scopedInstantiationService.invokeFunction(fn);
+	getInternalContextKeyService(): IContextKeyService {
+		return this.scopedInstantiationService.invokeFunction(accessor => accessor.get(IContextKeyService));
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/editor/editorGroupView.ts:865]: @roblourens why not keep the scoped context key service around to return directly? Could even just be a getter.

File: src/vs/workbench/browser/parts/editor/textEditor.ts [hunk 6]
```diff
@@ -200,6 +201,18 @@ export abstract class BaseTextEditor extends EditorPane implements ITextEditorPa
 		// Subclasses can override
 	}
 
+	getInternalContextKeyService(): IContextKeyService | undefined {
+		return this.invokeWithinContext(accessor => accessor.get(IContextKeyService));
+	}
+
+	private invokeWithinContext<T>(fn: (accessor: ServicesAccessor) => T): T | undefined {
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/editor/textEditor.ts:208]: I would merge this private method into the `getInternalContextKeyService` method.

File: src/vs/workbench/browser/parts/editor/textDiffEditor.ts [hunk 7]
```diff
@@ -60,6 +61,23 @@ export class TextDiffEditor extends BaseTextEditor implements ITextDiffEditorPan
 		this.doSaveOrClearTextDiffEditorViewState(editor);
 	}
 
+	getInternalContextKeyService(): IContextKeyService | undefined {
+		return this.invokeWithinDiffEditorContext(accessor => accessor.get(IContextKeyService));
+	}
+
+	private invokeWithinDiffEditorContext<T>(fn: (accessor: ServicesAccessor) => T): T | undefined {
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/editor/textDiffEditor.ts:68]: I would merge this private method into the getInternalContextKeyService method.

File: src/vs/workbench/services/editor/browser/editorService.ts [hunk 8]
```diff
@@ -780,20 +781,28 @@ export class EditorService extends Disposable implements EditorServiceImpl {
...
-			return activeTextEditorControl.invokeWithinContext(fn);
+	getEditorContextKeyService(): IContextKeyService {
+		const activeEditorPane = this.activeEditorPane;
+		if (activeEditorPane) {
+			const result = activeEditorPane.getInternalContextKeyService();
+			if (result) {
+				return result;
+			}
+
+			const activeControl = activeEditorPane.getControl();
```
Comment (bpasero) [inline src/vs/workbench/services/editor/browser/editorService.ts:794]: This now feels to me like it is at the wrong layer. Implementors of editors should do the right thing, I would expect the editor service to not have any smarts.

**Review by bpasero — CHANGES_REQUESTED** [review]
Besides my feedback inside code, some overall feedback on the naming and use of methods:
* `editorService.getEditorContextKeyService()` change this to a getter `editorService.activeEditorContextKeyService` (similar to `activeEditor`)
* `IEditorGroup.getInternalContextKeyService` change this to a getter `IEditorGroup.groupContextKeyService` or `IEditorGroup.scopedContextKeyService` (I find the name `internal` not pleasant)
* `IEditorPane.getInternalContextKeyService`: I am not sure this can be a getter given it is overridden by  subclasses but I would prefer a different name at least: `editorContextKeyService` or `scopedContextKeyService` as `internal` does not sound pleasant

**Comment by roblourens** [comment]
One difference now is that the EditorService's helper only returns the CKS for an active editor, so consumers like commandsQuickAccess have to do the fallback to the group CKS if it returns undefined. I think this is fine and expected though.

**Comment by bpasero** [comment]
Let me know when I should do another review.

**Comment by roblourens** [comment]
It's ready now

File: src/vs/workbench/browser/parts/editor/editorGroupView.ts [hunk 9]
```diff
@@ -869,8 +870,8 @@ export class EditorGroupView extends Themable implements IEditorGroupView {
 		}
 	}
 
-	invokeWithinContext<T>(fn: (accessor: ServicesAccessor) => T): T {
-		return this.scopedInstantiationService.invokeFunction(fn);
+	get scopedContextKeyService(): IContextKeyService {
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/editor/editorGroupView.ts:873]: Just :lipstick: but I prefer `get` and the associated `private` field to be close to each other, so can you move it up right below `_scopedContextKeyService`?

File: src/vs/workbench/services/editor/browser/editorService.ts [hunk 10]
```diff
@@ -428,6 +429,10 @@ export class EditorService extends Disposable implements EditorServiceImpl {
 		return activeCodeEditor?.getModel()?.getLanguageIdentifier().language;
 	}
 
+	get activeEditorContextKeyService(): IContextKeyService | undefined {
+		return this.activeEditorPane?.scopedContextKeyService;
```
Comment (bpasero) [inline src/vs/workbench/services/editor/browser/editorService.ts:433]: Shouldn't this fallback to use the scoped context key service of the active group? If you check how `invokeWithinEditorContext` used to be:

```
		const activeTextEditorControl = this.activeTextEditorControl;	
		if (isCodeEditor(activeTextEditorControl)) {	
			return activeTextEditorControl.invokeWithinContext(fn);	
		}	

		const activeGroup = this.editorGroupService.activeGroup;	
		if (activeGroup) {	
			return activeGroup.invokeWithinContext(fn);	
		}	

		return this.instantiationService.invokeFunction(fn);
```

If we end up not needing this anymore then I would actually think we can drop `activeEditorContextKeyService` altogether and let people use `activeEditorPane.scopedContextKeyService` directly.

File: src/vs/workbench/browser/parts/editor/textEditor.ts [hunk 11]
```diff
@@ -200,6 +201,14 @@ export abstract class BaseTextEditor extends EditorPane implements ITextEditorPa
 		// Subclasses can override
 	}
 
+	get scopedContextKeyService(): IContextKeyService | undefined {
+		if (!this.editorControl) {
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/editor/textEditor.ts:205]: You can remove this check for `undefined`, it is taken care of below.

File: src/vs/workbench/browser/parts/editor/textDiffEditor.ts [hunk 12]
```diff
@@ -60,6 +61,18 @@ export class TextDiffEditor extends BaseTextEditor implements ITextDiffEditorPan
...
+	get scopedContextKeyService(): IContextKeyService | undefined {
+		const control = this.getControl();
+		if (!control) {
+			return undefined;
+		}
+
+		const useEditor = control.getOriginalEditor().hasTextFocus() ?
+			control.getOriginalEditor() :
+			control.getModifiedEditor();
+		return useEditor.invokeWithinContext(accessor => accessor.get(IContextKeyService));
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/editor/textDiffEditor.ts:73]:
```suggestion
		return (control.getOriginalEditor().hasTextFocus() ?
			control.getOriginalEditor() :
			control.getModifiedEditor()).invokeWithinContext(accessor => accessor.get(IContextKeyService));
```

Comment (bpasero) [reply src/vs/workbench/browser/parts/editor/textDiffEditor.ts:73]: And I prefer `get` properties above constructor.

File: src/vs/workbench/electron-sandbox/window.ts [hunk 13]
```diff
@@ -457,7 +459,8 @@ export class NativeWindow extends Disposable {
 
 	private doUpdateTouchbarMenu(scheduler: RunOnceScheduler): void {
 		if (!this.touchBarMenu) {
-			this.touchBarMenu = this.editorService.invokeWithinEditorContext(accessor => this.menuService.createMenu(MenuId.TouchBarContext, accessor.get(IContextKeyService)));
+			const menuContextKeyService = this.editorService.activeEditorContextKeyService || this.editorGroupService.activeGroup.scopedContextKeyService;
```
Comment (bpasero) [inline src/vs/workbench/electron-sandbox/window.ts:462]: Again I think the fallback to the active group should be part of `this.editorService.activeEditorContextKeyService`

**Review by bpasero — CHANGES_REQUESTED** [review]
@roblourens I think we are getting close, some feedback provided, mainly polish.

One thing I might have missed is the idea of `this.editorService.activeEditorContextKeyService` falling back to the active group or not.

I think you implemented the latter that we no longer fallback, which I guess is fine if we make sure we use the group context key as needed (you probably checked that).

But if we do not fallback, I would suggest that people simply use `this.editorService.activeEditorPane.scopedContextKeyService`

**Comment by roblourens** [comment]
Is it beyond the scope of the editor service to do the fallback to the group context key service? I don't have a strong opinion but this needs a different name if it will have the fallback because I think it should match the other `activeEditor*` helpers if it is named like them. Really what you would be asking for is the "most relevant context key service right now".

**Comment by bpasero** [comment]
@roblourens yeah agree, so I am fine with your strategy as it is now, assuming that those places that need the fallback have it adopted.

**Comment by roblourens** [comment]
Yes, it's only used in 2 places and both have the fallback. All the feedback is handled, good to merge then?

**Comment by bpasero** [comment]
Its fine, but I pushed some changes in https://github.com/microsoft/vscode/pull/104947/commits/a59e709afd1ecc45e8d4bdb5f802612e25be7dd6 on top. I felt that `activeEditorPane.scopedContextKeyService` is good enough to have.
